osu.Game/Online/Rooms/OnlinePlayBeatmapAvailabilityTracker.cs

```diff
@@ -58,39 +64,50 @@ protected override void LoadComplete()
+                        beginTracking(beatmap);
+                }), TaskContinuationOptions.OnlyOnRanToCompletion);
+            }, true);
+        }
 
-                AddInternal(downloadTracker);
+        private void beginTracking(APIBeatmap beatmap)
+        {
+            Debug.Assert(beatmap.BeatmapSet != null);
 
-                downloadTracker.State.BindValueChanged(_ => Scheduler.AddOnce(updateAvailability), true);
-                downloadTracker.Progress.BindValueChanged(_ =>
-                {
-                    if (downloadTracker.State.Value != DownloadState.Downloading)
-                        return;
-
-                    // incoming progress changes are going to be at a very high rate.
-                    // we don't want to flood the network with this, so rate limit how often we send progress updates.
-                    if (progressUpdate?.Completed != false)
-                        progressUpdate = Scheduler.AddDelayed(updateAvailability, progressUpdate == null ? 0 : 500);
-                }, true);
-
-                // handles changes to hash that didn't occur from the import process (ie. a user editing the beatmap in the editor, somehow).
-                realmSubscription?.Dispose();
-                realmSubscription = realm.RegisterForNotifications(r => filteredBeatmaps(), (items, changes, ___) =>
-                {
-                    if (changes == null)
-                        return;
+            downloadTracker = new BeatmapDownloadTracker(beatmap.BeatmapSet);
+
+            AddInternal(downloadTracker);
 
-                    Scheduler.AddOnce(updateAvailability);
-                });
+            downloadTracker.State.BindValueChanged(_ => Scheduler.AddOnce(updateAvailability, beatmap), true);
+            downloadTracker.Progress.BindValueChanged(_ =>
+            {
+                if (downloadTracker.State.Value != DownloadState.Downloading)
+                    return;
+
+                // incoming progress changes are going to be at a very high rate.
+                // we don't want to flood the network with this, so rate limit how often we send progress updates.
+                if (progressUpdate?.Completed != false)
+                    progressUpdate = Scheduler.AddDelayed(updateAvailability, beatmap, progressUpdate == null ? 0 : 500);
             }, true);
+
+            // handles changes to hash that didn't occur from the import process (ie. a user editing the beatmap in the editor, somehow).
+            realmSubscription?.Dispose();
+            realmSubscription = realm.RegisterForNotifications(r => filteredBeatmaps(beatmap), (items, changes, ___) =>
+            {
+                if (changes == null)
+                    return;
+
+                Scheduler.AddOnce(updateAvailability, beatmap);
+            });
         }
 
-        private void updateAvailability()
+        private void updateAvailability(APIBeatmap beatmap)
```

Comment:
Is there any specific reason this isn't stored to a field instead?

Reply:
Hmm... It's alright if you'd prefer it that way, for me it feels unsafe from a race condition perspective (there are multiple levels of scheduling here). Trying to avoid `null` as much as I can and take a more functional approach in regards to data flow.